Choosing a Dashboard (#2428)
wpilibsuite/frc-docs: source/docs/software/dashboards/dashboard-intro.rst

:ref:`SmartDashboard <docs/software/dashboards/smartdashboard/index:SmartDashboard>` (Driver) - simple and efficient dashboard that uses relatively few computer resources.  It does not have the fancy look or some of the features Shuffleboard has, but it displays network tables data with a variety of widgets without bogging down the driver station computer.

:ref:`Shuffleboard <docs/software/dashboards/shuffleboard/index:Shuffleboard>` (Driver) - straightforward and easily customizable dashboard. It displays network tables data using a variety of widgets that can be positioned and controlled with robot code. It includes many extra features like: tabs, recording / playback, and advanced custom widgets.

> :ref:`Glass <docs/software/dashboards/glass/index:Glass>` (Programming) - robot data visualization tool. Its GUI is extremely similar to that of the :ref:`Simulation GUI <docs/software/wpilib-tools/robot-simulation/simulation-gui:Simulation Specific User Interface Elements>`. In its current state, it is meant to be used as a programmer's tool rather than a proper dashboard in a competition environment.

Reviewer: ```suggestion
:ref:`Glass <docs/software/dashboards/glass/index:Glass>` (Programming) - robot data visualization tool. Its GUI is extremely similar to that of the :ref:`Simulation GUI <docs/software/wpilib-tools/robot-simulation/simulation-gui:Simulation Specific User Interface Elements>`. In its current state, it is meant to be used as a programmer's tool rather than a proper dashboard in a competition environment, with a focus on high performance real time plotting.
```

> :ref:`AdvantageScope <docs/software/dashboards/glass/index:Glass>` (Programming) - robot diagnostics, log review/analysis, and data visualization application.  It reads logs in WPILOG, DS log, and RLOG file formats, plus live robot data viewing using NT4 or RLOG streaming.

Author: Correct.

This was shamelessly stolen from the [AdvantageScope page](https://github.com/Mechanical-Advantage/AdvantageScope/tree/fdda52f045c2953d5c3c9e3dd43b2b81ffdb634a)

I believe the file extensions are: .wpilog, .dslog, and .rlog

I don't know anything about rlog.  I guess the question is do we indicate the extension as all caps as shown or `.wpilog`.  I have seen it both ways.

@jwbonner

> :ref:`AdvantageScope <docs/software/dashboards/glass/index:Glass>` (Programming) - robot diagnostics, log review/analysis, and data visualization application.  It reads logs in WPILOG, DS log, and RLOG file formats, plus live robot data viewing using NT4 or RLOG streaming.

Reviewer 2: I think most places in the docs currently refer to it as a “WPILib Data Log” instead of using an abbreviation. I’ve also seen it as WPILog. If we standardize on capitalization/formatting I’m happy to update the AdvantageScope docs too for consistency.

RLOG appears in all caps everywhere in our docs. But honestly I think that should just be removed here - it’s a legacy format used by AdvantageKit in 2022, and we’ve kept it for backwards compatibility with old logs. New users will never have a use for it.

> :ref:`AdvantageScope <docs/software/dashboards/glass/index:Glass>` (Programming) - robot diagnostics, log review/analysis, and data visualization application.  It reads logs in WPILOG, DS log, and RLOG file formats, plus live robot data viewing using NT4 or RLOG streaming.

Author: This would be my preference.  Objections?

"It reads the WPILib Data Log (``.wpilog``) and Driver Station Log (``.dslog``) file formats, plus live robot data viewing using NT4."

> :ref:`AdvantageScope <docs/software/dashboards/glass/index:Glass>` (Programming) - robot diagnostics, log review/analysis, and data visualization application.  It reads logs in WPILOG, DS log, and RLOG file formats, plus live robot data viewing using NT4 or RLOG streaming.

Reviewer 2: I would just amend it to something like "Driver Station Log (`.dslog`/`.dsevents`)" since the DS console log is supported too. Otherwise it looks good to me.

> :ref:`AdvantageScope <docs/software/dashboards/glass/index:Glass>` (Programming) - robot diagnostics, log review/analysis, and data visualization application.  It reads the WPILib Data Log (``.wpilog``) and Driver Station Log (``.dslog`` / ``.dsevents``) file formats, plus live robot data viewing using NT4.

Reviewer: ```suggestion
:ref:`AdvantageScope <docs/software/dashboards/glass/index:Glass>` (Programming) - robot diagnostics, log review/analysis, and data visualization application.  It reads the WPILib Data Log (``.wpilog``) and Driver Station Log (``.dslog`` / ``.dsevents``) file formats, plus live robot data viewing.
```
No need to mention NT4 (applies to effectively all dashboards).

Third Party Dashboards
----------------------

`FRC Web Components <https://github.com/frc-web-components/frc-web-components>`__ (Driver) - A web-based dashboard that can be installed as a standalone application, or as a JavaScript package for custom dashboard solutions.

> `Elastic <https://github.com/Gold872/elastic-dashboard>`__ (Driver) - simple and modern Shuffleboard alternative made by Team 353. It is meant to serve as a dashboard for competition but can also be used for testing.  Featuring draggable and resizable card widgets and automatic IP retrieval from the FRC Driver Station

Reviewer: ```suggestion
`Elastic <https://github.com/Gold872/elastic-dashboard>`__ (Driver) - simple and modern Shuffleboard alternative made by Team 353. It is meant to serve as a dashboard for competition but can also be used for testing.  It features draggable and resizable card widgets.
```
Basically all dashboards get their IP from the DS.

> `QFRCDashboard <https://github.com/binex-dsk/QFRCDashboard>`__ (Driver) - reliable, high-performance, low-footprint dashboard.  QFRCDashboard has been specifically designed to use up as little resources as possible.

Reviewer 3: Reliable might overselling it, since it was just released and hasn't had much use.

Author: I copied that from their page.  Not sure how to resolve that.  Maybe say like "described as" or something.  Reliability is in the eye of the beholder, but I agree that it is near impossible to make that claim with as little testing as it has likely had when compared with say shuffleboard.

Author: I didn't think the others needed that label so just added that one

> `QFRCDashboard <https://github.com/binex-dsk/QFRCDashboard>`__ (Driver) - described as reliable, high-performance, low-footprint dashboard.  QFRCDashboard has been specifically designed to use up as little resources as possible.

Reviewer 4: "use up as little resources as possible" sounds odd to me, especially with a renewable resource like RAM and CPU. I think "use as few resources as possible." may be better?

Author: I agree.  I stole it from their site but that is better grammar

Choosing a Dashboard
====================

A dashboard is a program used to retrieve and display information about the operation of your robot.  There are two main types of dashboards that teams may need: driver and programmer dashboards.  Some dashboards will try to accommodate both purposes.

Driver Dashboard
----------------

During competition the drive team will use this dashboard to get information from the robot.  It should focus on conveying key information instantly.  This is often best accomplished by using large, colorful, and easy to understand visual elements.

> Most teams will use this dashboard to select their autonomous routine.  It is often helpful to hide this choice after the selection is made as that information is not needed during the match.

Reviewer 3: The part about hiding the autonomous selection seems out of place. I'm also not sure how to do it except in Shuffleboard by selecting another tab. Recommend removing from here, and possibly adding to individual dashboard articles with more details if applicable.

Programming Dashboard
---------------------

This dashboard is designed for debugging code and analyzing data from the robot. It supports the monitoring of a wide variety of information simultaneously, prioritizing function and utility over simplicity or ease of use. This functionality often includes complex data visualization and graphing across extended periods. In scenarios where there is an overwhelming amount of data to review, real-time analysis becomes challenging. The capability to examine past data and replay it proves to be extremely beneficial. While some dashboards may log data transmitted to them, :ref:`on-robot telemetry <docs/software/telemetry/datalog:On-Robot Telemetry Recording Into Data Logs>` using the ``Datalog`` class simplifies the process.

Specific Dashboards (oldest to newest)
--------------------------------------

> .. note:: SmartDashboard and Shuffleboard have a long history of aiding FRC teams. However, they are not currently receiving active updates or improvements. Notably, Shuffleboard may experience performance issues on some machines under certain scenarios.

Reviewer 3: I think I'd say something like Smatdashboard and Shuffleboard do no have a current WPILib maintainer, so no new features will be added or invasive bugs fixed. PRs from external contributors will be reviewed.
